src/freetype/ft-ogl.adb

```diff
@@ -0,0 +1,285 @@
+      Texture_2D.Set_Y_Wrapping (GL.Objects.Textures.Clamp_To_Edge); --  Wrap_T
+
+      if Width > 0 and then Height > 0 then
+         Texture_2D.Storage (Num_Levels, RGBA8, Width, Height);
+      else
+         Texture_2D.Storage (Num_Levels, RGBA8, 1, 1);
+      end if;
+
+      Bitmap_Image_Ptr :=  Bitmap.Buffer;
+      Texture_2D.Load_Sub_Image_From_Data
+        (Mip_Level_0, X_Offset, Y_Offset, Width, Height, Red, Unsigned_Byte,
+         Bitmap_Image_Ptr);
+      Char_Data.Texture := aTexture;
+   end Load_Texture;
+
+   -- --------------------------------------------------------------------------
+
+   procedure Print_Character_Metadata (Data : Character_Record) is
+      use GL.Types;
+      use FT.Faces;
+   begin
+      Put_Line ("Width: " & GL.Types.Int'Image (Data.Width));
+      Put_Line ("Rows: " & GL.Types.Int'Image (Data.Rows));
+      Put_Line ("Left: " & GL.Types.Int'Image (Data.Left));
+      Put_Line ("Top: " & GL.Types.Int'Image (Data.Top));
+      Put_Line ("Advance X: " & GL.Types.Int'Image (Advance_X (Data)) & " bits");
+      New_Line;
+   end Print_Character_Metadata;
+
+   --  ------------------------------------------------------------------------
+
+   procedure Render_Text (Render_Program : GL.Objects.Programs.Program;
+                          Text   : String; X, Y, Scale : GL.Types.Single;
+                          Colour : GL.Types.Colors.Basic_Color;
+                          Texture_ID, Projection_Matrix_ID, Colour_ID : GL.Uniforms.Uniform;
+                          Projection_Matrix : GL.Types.Singles.Matrix4) is
+      use GL.Objects.Buffers;
+      use GL.Objects.Textures.Targets;
+      use GL.Types.Colors;
+      use GL.Types;
+      use FT.Faces;
+
+      Num_Triangles  : constant GL.Types.Int := 2;
+      Num_Vertices   : constant GL.Types.Int := Num_Triangles * 3; -- Two triangles
+      Num_Components : constant GL.Types.Int := 4;                 -- Coords vector size;
+      Stride         : constant GL.Types.Int := 0;
+
+      Char           : Character;
+      Char_Data      : Character_Record;
+      Char_Texture   : GL.Objects.Textures.Texture;
+      X_Orig         : Single := X;
+      Y_Orig         : constant Single := Y;
+      X_Pos          : Single;
+      Y_Pos          : Single;
+      Char_Width     : Single;
+      Height         : Single;
+      --  2D quad as two triangles requires 2 * 3 vertices of 4 floats
+      Vertex_Data    : Singles.Vector4_Array (1 .. Num_Vertices);
+   begin
+      --  Blending allows a fragment colour's alpha value to control the resulting
+      --  colour which will be transparent for all the glyph's background colours and
+      --  non-transparent for the actual character pixels.
+      GL.Toggles.Enable (GL.Toggles.Blend);
+      GL.Blending.Set_Blend_Func (GL.Blending.Src_Alpha,
+                                  GL.Blending.One_Minus_Src_Alpha);
+      GL.Objects.Programs.Use_Program (Render_Program);
```

Comment:
Since the shaders are not part of this library (only provided for examples), I don't think it is a good idea to have program setup code in here. Users should setup an appropriate program for themselves. OTOH, if the exact shaders in your example are necessary for this package to do anything useful, they should be part of the main sources instead of the examples. Can you give a pointer about how vital the correct shaders are for this to work, and how much sense it would make using user-provided shaders?

Reply:
I had been thinking about moving the shaders and associated setup operations into OGL so that, for rendering text, the user would not have to know or do anything about the situation.
However, a problem is the the shaders are machine dependent in that the GLSL version number appropriate to the computer that they are being used on.
My feeling is that these shaders are specific and to OGL and the best thing would be that the user did not have to provide them.
Although of given this some thought, I haven't come to any conclusion.
With regard to the blending function, I found it essential to set and reset it for text rendering use as if permanently set it causes flickering problems with other graphics.
However, I've just realised that I should record the blend state on entry and return it to that state before leaving the routine.
More thought needed on my part.
With this I'll appreciate any suggestions you can provide.